DFC.App.DiscoverSkillsCareers/Controllers/AssessmentController.cs

```diff
@@ -1,88 +1,134 @@
+            {
+                return BadRequest();
+            }
+
+            if (!HasSessionId())
+            {
+                return RedirectToRoot();
+            }
+
+            var questionResponse = await GetQuestion(requestViewModel.AssessmentType, requestViewModel.QuestionNumber).ConfigureAwait(false);
+
+            if (questionResponse == null)
+            {
+                return BadRequest();
+            }
+
+            if (requestViewModel.QuestionNumber > questionResponse.MaxQuestionsCount)
             {
                 return BadRequest();
             }
 
-            var result = CreateResponseViewModel(viewModel.QuestionSetName, viewModel.QuestionId);
-            return View(result);
+            if (questionResponse.IsComplete)
+            {
+                return RedirectTo("results");
+            }
+
+            if (requestViewModel.QuestionNumber > questionResponse.QuestionNumber)
+            {
+                return RedirectTo($"assessment/{requestViewModel.AssessmentType}/{questionResponse.QuestionNumber}");
+            }
+
+            var responseViewModel = mapper.Map<QuestionGetResponseViewModel>(questionResponse);
+            return View(responseViewModel);
         }
 
         [HttpPost]
-        public IActionResult Index(QuestionPostRequestViewModel viewModel)
+        public async Task<IActionResult> Index(QuestionPostRequestViewModel requestViewModel)
         {
-            if (viewModel == null)
+            if (requestViewModel == null)
+            {
+                return BadRequest();
+            }
+
+            if (!HasSessionId())
+            {
+                return RedirectToRoot();
+            }
+
+            var question = await GetQuestion(requestViewModel.AssessmentType, requestViewModel.QuestionNumber).ConfigureAwait(false);
+            if (question == null)
             {
                 return BadRequest();
             }
 
-            var result = CreateResponseViewModel(viewModel.QuestionSetName, viewModel.QuestionId);
+            var result = mapper.Map<QuestionGetResponseViewModel>(question);
 
             if (!ModelState.IsValid)
             {
                 return View(result);
             }
 
-            if (string.IsNullOrWhiteSpace(result.NextQuestionId))
+            var answerResponse = await apiService.AnswerQuestion(requestViewModel.AssessmentType, requestViewModel.QuestionNumber, requestViewModel.Answer).ConfigureAwait(false);
+
+            if (answerResponse.IsSuccess)
             {
-                return Redirect($"assessment/complete");
+                if (answerResponse.IsComplete)
+                {
+                    return RedirectTo("assessment/complete");
+                }
+                else
+                {
+                    var assessmentTypeName = GetAssessmentTypeName(requestViewModel.AssessmentType);
+                    return RedirectTo($"assessment/{assessmentTypeName}/{answerResponse.NextQuestionNumber}");
```

Comment:
Wouldn't it be better to use redirecttoaction?

Reply:
Had issues with sonar cloud as it wont allow redirection based on user input so had to introduce redirectTo method